flocker/control/_persistence.py

```diff
@@ -164,6 +164,32 @@ def upgrade_from_v3(cls, config):
         }
         return dumps(decoded_config)
 
+    @classmethod
+    def upgrade_from_v4(cls, config):
+        """
+        Migrate a v4 JSON configuration to v5.
+
+        :param bytes config: The v4 JSON data.
+        :return bytes: The v5 JSON data.
+        """
+        decoded_config = loads(config)
+        decoded_config[u"version"] = 5
+        try:
+            nodes = decoded_config[u"deployment"][u"nodes"]
+        except KeyError:
+            pass
+        else:
+            new_nodes = []
+            for n in nodes:
+                new_node = n
+                new_node[u"applications"] = {
+                    u"values": [(a[u"name"], a) for a in n[u"applications"]],
+                    _CLASS_MARKER: "PMap"
```

Comment:
Why isn't `values` a dictionary? `dict((a.name, a) for a in n[u"applications"])`